Mono incorrectly throws a TypeLoadException for a recursive type definition

### Description

Given the following code:

```csharp
var test = new Node<int>();
Console.WriteLine(test);

public class Node<T>
{
  public Node<T>[][] Children;
  public T Data;
}
```
Mono will throw a `TypeLoadException` exception, which is not expected:
```
Unhandled Exception:
System.TypeLoadException: Recursive type definition detected .Node`1
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Recursive type definition detected .Node`1
```
CoreCLR does not throw a similar exception.

### Reproduction Steps

1. Create a new console app with the code above.
2. In the .csproj file, add the following content to run the console app with Mono:

```        
<SelfContained>true</SelfContained>
<UseMonoRuntime>true</UseMonoRuntime>
<RuntimeIdentifier>osx-x64</RuntimeIdentifier>
```
(Note that I'm using macOS x64, but the behavior is not specific to platform or architecture.)

3. Run the built console app, the exception will occur, which is not expected.

### Expected behavior

The output is:

```
Node`1[System.Int32]
```
 and no exception occurs.

### Actual behavior

Mono will throw a `TypeLoadException` exception, which is not expected:
```
Unhandled Exception:
System.TypeLoadException: Recursive type definition detected .Node`1
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeLoadException: Recursive type definition detected .Node`1
```

### Regression?

This has not worked in Mono. I also see the same behavior with Mono 6.12.0 from the mono/mono repository.

### Known Workarounds

I don't have any known worarounds.

### Configuration

I'm using .NET version 7.0.202 on macOS x64.

### Other information

This bug was originally reported to Unity here: https://issuetracker.unity3d.com/issues/typeloadexception-is-thrown-when-using-recursive-types-in-monobehaviour-inherited-classes

--- a/src/mono/mono/metadata/class-init.c
+++ b/src/mono/mono/metadata/class-init.c
@@ -1183,9 +1183,10 @@ mono_class_create_bounded_array (MonoClass *eclass, guint32 rank, gboolean bound
 
 	mono_class_setup_supertypes (klass);
 
-	//if (mono_class_is_ginst (eclass))
-	//	mono_class_init_internal (eclass);
-	// NOTE: this is also probably too aggressive if eclass is not a valuetype
+	// NOTE: this is probably too aggressive if eclass is not a valuetype.  It looks like we
+	// only need the size info in order to set MonoClass:has_references for this array type -
+	// and for that we only need to setup the fields of the element type if it's not a reference
+	// type.
 	if (!eclass->size_inited)
 		mono_class_setup_fields (eclass);
 	mono_class_set_type_load_failure_causedby_class (klass, eclass, "Could not load array element type");

--- a/src/mono/mono/metadata/class-init.c
+++ b/src/mono/mono/metadata/class-init.c
@@ -1183,8 +1183,9 @@ mono_class_create_bounded_array (MonoClass *eclass, guint32 rank, gboolean bound
 
 	mono_class_setup_supertypes (klass);
 
-	if (mono_class_is_ginst (eclass))
-		mono_class_init_internal (eclass);
+	//if (mono_class_is_ginst (eclass))
+	//	mono_class_init_internal (eclass);
+	// NOTE: this is also probably too aggressive if eclass is not a valuetype
 	if (!eclass->size_inited)
 		mono_class_setup_fields (eclass);
 	mono_class_set_type_load_failure_causedby_class (klass, eclass, "Could not load array element type");

--- a/src/tests/Loader/classloader/regressions/GitHub_85821/GitHub_85821.csproj
+++ b/src/tests/Loader/classloader/regressions/GitHub_85821/GitHub_85821.csproj
@@ -1,0 +1,9 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="repro.cs" />
+  </ItemGroup>
+</Project>

--- a/src/tests/Loader/classloader/regressions/GitHub_85821/repro.cs
+++ b/src/tests/Loader/classloader/regressions/GitHub_85821/repro.cs
@@ -12,11 +12,21 @@ public class Program
         var fit = test.GetField("Children").FieldType;
         if (fit == null)
             return 101;
-        
+
         var test2 = typeof(Node2<int>);
         var fit2 = test2.GetField("Children").FieldType;
         if (fit2 == null)
             return 102;
+
+        var test3 = typeof(NodeVT<double>);
+        var fit3 = test3.GetField("Children").FieldType;
+        if (fit3 == null)
+            return 103;
+
+        var test4 = typeof(NodeVT2<int>);
+        var fit4 = test4.GetField("Children").FieldType;
+        if (fit4 == null)
+            return 104;
 
         return 100;
     }
@@ -27,8 +37,17 @@ public class Node<T>
     public Node<T>[] Children;
 }
 
-
 public class Node2<T>
 {
     public Node2<T>[][] Children;
+}
+
+public struct NodeVT<T>
+{
+    public NodeVT<T>[] Children;
+}
+
+public struct NodeVT2<T>
+{
+    public NodeVT2<T>[][] Children;
 }

--- a/src/tests/Loader/classloader/regressions/GitHub_85821/repro.cs
+++ b/src/tests/Loader/classloader/regressions/GitHub_85821/repro.cs
@@ -1,0 +1,34 @@
+using System;
+
+/* Regression test for https://github.com/dotnet/runtime/issues/85821
+ * ensure that self-referencing generic instances are initialized correctly and don't TLE
+ */
+
+public class Program
+{
+    public static int Main()
+    {
+        var test = typeof(Node<double>);
+        var fit = test.GetField("Children").FieldType;
+        if (fit == null)
+            return 101;
+        
+        var test2 = typeof(Node2<int>);
+        var fit2 = test2.GetField("Children").FieldType;
+        if (fit2 == null)
+            return 102;
+
+        return 100;
+    }
+}
+
+public class Node<T>
+{
+    public Node<T>[] Children;
+}
+
+
+public class Node2<T>
+{
+    public Node2<T>[][] Children;
+}

